### PR TITLE
Improve efficiency

### PR DESCRIPTION
**Pull request by hyeok9855**
- [x] I've read the [.github/CONTRIBUTING.md](.github/CONTRIBUTING.md) file
- [x] My code follows the typing guidelines
- [x] I've added appropriate tests
- [x] I've run pre-commit hooks locally

## Description
Improve code efficiency by:
1) making action validity check optional
2) calculating rewards outside the sampling loop
3) and several minor improvements.

When tested using `python -O tutorials/examples/train_hypergrid_simple.py --no_cuda --ndim 2 --height 8` the runtime reduced **from 20s to 15s**.

### CODE REVIEW

File: src/gfn/env.py [hunk 1]
```diff
@@ -285,43 +287,41 @@ def _step(self, states: States, actions: Actions) -> States:
...
+                )
+
+        new_valid_states_idx = valid_states_idx & ~actions.is_exit
 
         # IMPORTANT: states.clone() is used to ensure that the new states are a
         # distinct object from the old states. This is important for the sampler to
         # work correctly when building the trajectories. If you want to override this
         # method in your custom environment, you must ensure that the `new_states`
         # returned is a distinct object from the submitted states.
-        new_states = states.clone()
```
[the remark this answers is not in the document]
Comment (hyeok9855) [reply src/gfn/env.py:294]: I hoped so, but I guess it isn't. Let me roll back the change.

File: src/gfn/env.py [hunk 2]
```diff
@@ -285,43 +287,41 @@ def _step(self, states: States, actions: Actions) -> States:
...
+        ), f"The step function must return a States instance, but got {type(not_done_states)} instead."
 
-        # Set to the sink state when the action is exit.
-        new_sink_states_idx = actions.is_exit
-        sf_states = self.States.make_sink_states(
-            (int(new_sink_states_idx.sum().item()),), device=states.device
+        new_states = self.States.make_sink_states(
+            states.batch_shape, device=states.device
         )
-        new_states[new_sink_states_idx] = sf_states
```
[the remark this answers is not in the document]
Comment (hyeok9855) [reply src/gfn/env.py:312]: Sorry for the inconvenience, and thanks for pointing it out! I totally agree and will add some comments.

File: src/gfn/estimators.py [hunk 3]
```diff
@@ -307,11 +309,7 @@ def to_probability_distribution(
         probs = torch.softmax(logits, dim=-1)
 
         if epsilon != 0.0:
-            uniform_dist_probs = torch.where(
-                masks.sum(dim=-1, keepdim=True) == 0,
-                torch.zeros_like(masks),
-                masks.to(torch.get_default_dtype()) / masks.sum(dim=-1, keepdim=True),
-            )
+            uniform_dist_probs = masks / masks.sum(dim=-1, keepdim=True)
```
[the remark this answers is not in the document]
Comment (hyeok9855) [reply src/gfn/estimators.py:312]: Here, dividing with `masks.sum(dim=-1, keepdim=True)` automatically converts the output dtype to the default dtype.

The output should be the same as before! The reason we used `torch.where` before was to handle the case where all actions are masked, but this shouldn't happen here in the estimator, and that's why I added the assertion above.

**Comment by hyeok9855** [comment]
@josephdviviano 
When tested using `python tutorials/examples/train_hypergrid_simple.py --ndim 2 --height 32 --n_iterations 2000 --no_cuda`,

- Before this PR: ~46 sec
- After this PR: ~40 sec (~13% faster)

When tested with the additional `-O` (which sets `__debug__==True` and skips all assertions) flag,

- Before this PR: ~41 sec
- After this PR: ~35 sec (~15% faster)

All tested on my MBP (M4PRO).

File: src/gfn/env.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (josephdviviano) [reply src/gfn/env.py:294]: This was the cause of an extremely nasty bug. So I'll move the comment. We need to ensure anyone who subclasses `env.py` does not fail to clone the state.

File: src/gfn/env.py [hunk 4]
```diff
@@ -285,43 +287,41 @@ def _step(self, states: States, actions: Actions) -> States:
...
+        not_done_states = self.step(not_done_states, not_done_actions)
+        assert isinstance(
+            not_done_states, States
+        ), f"The step function must return a States instance, but got {type(not_done_states)} instead."
 
-        # Set to the sink state when the action is exit.
-        new_sink_states_idx = actions.is_exit
-        sf_states = self.States.make_sink_states(
-            (int(new_sink_states_idx.sum().item()),), device=states.device
+        new_states = self.States.make_sink_states(
```
[the remark this answers is not in the document]
Comment (josephdviviano) [reply src/gfn/env.py:328]: Did you add any comments?

File: src/gfn/env.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (josephdviviano) [reply src/gfn/env.py:312]: Thanks your comments are excellent. It's really no problem, I just want to ensure the library is as user-friendly as possible, and I appreciate your attention to detail.
